File: June-LeetCoding-Challenge-2020/Week1/518-coin-change-2.py
class Solution:
    # Sol 1. DP
    def change(self, amount: int, coins: List[int]) -> int:
        dp = [0] * (amount + 1)
        dp[0] = 1
        #print(dp)
        for coin in coins:
            #print(coin)
            for x in range(coin, amount + 1):
                dp[x] += dp[x - coin]
                #print(dp)
        return dp[amount]
    # TC: O(nm), where n is the length of coins, m is amount 
    # SC: O(m+1)
    # Runtime: 136 ms, faster than 90.14%
    # Memory Usage: 14 MB, less than 65.82%.

File: June-LeetCoding-Challenge-2020/Week1/518-coin-change-2.cpp
// Time Complexity: O(coins count * amount)
// Space Complexity: O(amount + 1)
// Runtime: 12 ms, faster than 71.21%
// Memory Usage: 6.9 MB, less than 80.65%
class Solution {
public:
    int change(int amount, vector<int>& coins) {
        
        // amount 0 as initializer + [1,amount]
        int number_of_ways[amount+1];
        number_of_ways[0] = 1;
        
        // initialize amounts to 0 
        for(int i = 1; i <= amount; i++){
            number_of_ways[i] = 0;
        }
        
        for(int coin_index = 0; coin_index < coins.size(); coin_index++){
            for(int target_amount = 1; target_amount <= amount; target_amount++){
                if(target_amount >= coins[coin_index]){
                    number_of_ways[target_amount] += number_of_ways[target_amount - coins[coin_index]];
                }
            }
        }
        
        return number_of_ways[amount];
    }
};